File: hackerrank/30 days of code/day 23, bst level-order traversal/day_23_bst_level_order_traversal.rb
class Node
    attr_accessor :left,:right,:data
  def initialize data
      @left=@right=nil
      @data=data
    end
end
class Solution
    def insert (root,data)
        if root==nil
            return Node.new(data)
        else
            if data<=root.data
                cur=self.insert(root.left,data)
                root.left=cur
            else
                cur=self.insert(root.right,data)
                root.right=cur
            end
        end
        return root
    end

    # Immutable HackerRank code above
    def levelOrder(root)

        if not root
            return
        end
        
        print root.data, " "

        myQueue = Queue.new
        myQueue << root

        while not myQueue.empty?

            first = myQueue.pop
            if first.left
                myQueue << first.left
                print first.left.data, " "
            end
            if first.right
                myQueue << first.right
                print first.right.data, " "
            end
        end
    end
end

# Immutable HackerRank code below
myTree=Solution.new
root=nil
T=gets.to_i
for i in 1..T
    data=gets.to_i
    root=myTree.insert(root,data)
end
myTree.levelOrder(root)

File: hackerrank/30 days of code/day 23, bst level-order traversal/day_23_bst_level_order_traversal.py
import sys


class Node:
    def __init__(self, data):
        self.right = self.left = None
        self.data = data


class Solution:
    def insert(self, root, data):
        if root == None:
            return Node(data)
        else:
            if data <= root.data:
                cur = self.insert(root.left, data)
                root.left = cur
            else:
                cur = self.insert(root.right, data)
                root.right = cur
        return root

    # Immutable HackerRank code above
    def levelOrder(self, root):

        from collections import deque

        if not root:
            return

        myQueue = deque()
        myQueue.append(root)
        print(myQueue[0].data, end=" ")

        while len(myQueue):

            if myQueue[0].left:
                myQueue.append(myQueue[0].left)
                print(myQueue[0].left.data, end=" ")

            if myQueue[0].right:
                myQueue.append(myQueue[0].right)
                print(myQueue[0].right.data, end=" ")

            myQueue.popleft()


# Immutable HackerRank code below
T = int(input())
myTree = Solution()
root = None
for i in range(T):
    data = int(input())
    root = myTree.insert(root, data)
myTree.levelOrder(root)
